### ISSUE TITLE
after traning how to create models?

### ISSUE THREAD

**Issue by UlasSAYGINIM** [Calamari-OCR/calamari#215]
I run train command and it saved checkpoints in output directory. 
now , how can i use these as model? how to make models?
which command?
is there any more detailed documentation for this?

**Comment by ChWick** [comment]
Either call `calamari-predict` (see [here](https://github.com/Calamari-OCR/calamari#prediction-of-a-page)) or use the [Python-API](https://github.com/Calamari-OCR/calamari/wiki/Predicting-(Python-API)).

**Comment by UlasSAYGINIM** [comment]
i need to code in python? there is no command like 

makemodel --checkpoing folderpath --output newmodels

**Comment by UlasSAYGINIM** [comment]
@ChWick thank you for you help.

**Comment by ChWick** [comment]
If `best.cpkt` and `best.ckpt.json` are present you successfully trained a model that is ready to use with `calamari-predict` *or* the python API.

The python API is optional, just use the command line scripts if you prefer to!

**Comment by UlasSAYGINIM** [comment]
if there is no json that means i need to train more. thank you.

but it say something like this
WARNING  2021-03-20 14:26:41,292 tfaip.base.scenario.scenarioba: Training without validation.

what does it mean?

and do i need to use only one line text to train? or may i use paragraphs?? because when i use paragraphs it stops at epoch 1 and no any improvement. it waits all day :)

**Comment by UlasSAYGINIM** [comment]
how i can use gpu for training i did cuda settings it sees gpu and onlu uses cpu and cpu getting hot.

**Comment by ChWick** [comment]
Ah, now I understand. See this issue [here](https://github.com/Calamari-OCR/calamari/issues/207#issuecomment-781364194)
You need a validation list or pass a 'train' split value to obtain the 'best' model.

**Comment by ChWick** [comment]
Update calamari to 2.0.2 (was just released) that uses `--use_train_as_val` by default (`--validation_split_ratio 0.2` might be preferred though)

**Comment by UlasSAYGINIM** [comment]
Thanks a lot @ChWick , i will update and try again.

you said that 

> You need a validation list or pass a 'train' split value to obtain the 'best' model.

how is the validation list format? and splitting value?

for example, i am putting image which written "abcdhefg" than i write same thing gt.txt.

is that right form of data to provide training?

**Comment by ChWick** [comment]
assume you files are labelled:  `0001.png` (the image) and `0001.gt.txt` (the Ground Truth of the image, i.e. the written text), and you have 1000 lines, so the last pair is `1000.png` and `1000.gt.txt`. You could now either manually split these files (randomly) into two directories train and val to obtain a split, i.e. `0001` up to `0200` in val and `0201` up to `1000` in train. Then pass `--files train/*.png` and `--validation val/*.png`. Alternatively, use the `--validation_split_ratio 0.2 --files *.png` (all files). then Calamari will split your data into train and val automatically.

**Comment by UlasSAYGINIM** [comment]
i tried to setup latest version but when i started install it installs old version of numpy 1.1.8.5 and tensorflow-estimator 2.3.0 and than it gives incompatible versions error like below.

> Installing collected packages: numpy, tensorflow-estimator
  WARNING: The script f2py.exe is installed in 'C:\Program Files\Python38\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow-gpu 2.4.1 requires numpy~=1.19.2, but you have numpy 1.18.5 which is incompatible.
tensorflow-gpu 2.4.1 requires tensorflow-estimator<2.5.0,>=2.4.0, but you have tensorflow-estimator 2.3.0 which is incompatible.
Successfully installed numpy-1.18.5 tensorflow-estimator-2.3.0


even i uninstall these incompatible versions than install again. it again installs old version of them and give same error :(

i have multiple python versions. and  because of that i deleted environment path entries. not to see any conflicts about versions.

**Comment by ChWick** [comment]
Can you try tensorflow 2.3.1?

**Comment by UlasSAYGINIM** [comment]
i can try but when i typd these but nothing changed.

py -3.8 -m pip uninstall numpy
py -3.8 -m pip  install numpy==1.19.2

py -3.8 -m pip uninstall tensorflow-estimator

py -3.8 -m pip  install tensorflow-estimator==2.4.0

same error.
i will try again with environment path.

> tensorflow-gpu 2.4.1 requires tensorflow-estimator<2.5.0,>=2.4.0, but you have tensorflow-estimator 2.3.0 which is incompatible.

**Comment by ChWick** [comment]
i mean `pip remove tensorflow tensorflow-estimator numpy` and `pip install tensorflow==2.3.1`

**Comment by UlasSAYGINIM** [comment]
py -3.8 -m pip  install tensorflow-estimator==2.3.1
ERROR: Could not find a version that satisfies the requirement tensorflow-estimator==2.3.1
ERROR: No matching distribution found for tensorflow-estimator==2.3.1

**Comment by ChWick** [comment]
`py -3.8 -m pip install tensorflow==2.3.1` (remove the estimator!)

**Comment by UlasSAYGINIM** [comment]
i am doing what you said now.

**Comment by UlasSAYGINIM** [comment]
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow-gpu 2.4.1 requires numpy~=1.19.2, but you have numpy 1.18.5 which is incompatible.
tensorflow-gpu 2.4.1 requires tensorflow-estimator<2.5.0,>=2.4.0, but you have tensorflow-estimator 2.3.0 which is incompatible.

**Comment by UlasSAYGINIM** [comment]
Installing collected packages: numpy, tensorflow-estimator, tensorflow
  WARNING: The script f2py.exe is installed in 'C:\Program Files\Python38\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The scripts estimator_ckpt_converter.exe, saved_model_cli.exe, tensorboard.exe, tf_upgrade_v2.exe, tflite_convert.exe, toco.exe and toco_from_protos.exe are installed in 'C:\Program Files\Python38\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow-gpu 2.4.1 requires numpy~=1.19.2, but you have numpy 1.18.5 which is incompatible.
tensorflow-gpu 2.4.1 requires tensorflow-estimator<2.5.0,>=2.4.0, but you have tensorflow-estimator 2.3.0 which is incompatible.
Successfully installed numpy-1.18.5 tensorflow-2.3.1 tensorflow-estimator-2.3.0

**Comment by UlasSAYGINIM** [comment]
now py -3.8 -m pip install calamari_ocr works without problem.

**Comment by UlasSAYGINIM** [comment]
but it is not in the environment in named v1. i think it is installed global system.

**Comment by ChWick** [comment]
I think you falsely installed `tensorflow-gpu`, which should not be installed at all. The `tensorflow`-package (since 2.0) supports both cpu and gpu. Mayby start with a clean environment, if there are further problems.

**Comment by UlasSAYGINIM** [comment]
i added user path. now it seems everything fine.
i think problem about multiple python version installation. i deleted all system environment paths to prevent conflicts from python versions. now i am able to call calamari-predict and i will try now.

**Comment by ChWick** [comment]
Great!

**Comment by UlasSAYGINIM** [comment]
i used calamari-predict for turkish language but it seems model of uw3-modern-english does not able to recognize same turkish characters like ü , it predict that as double "i" like "ii".

how can i add only turkish characters for uw3-modern-english? how many png i should train for it?

**Comment by UlasSAYGINIM** [comment]
i accidently closed issue sorry. my mistake.

**Comment by ChWick** [comment]
Oups, I clearly missed that new question at the end of the discussion... sorry for that....

The number of training data highly depends on the quality of your data and the accuracy you want to achieve. Maybe try about 500 - 1000 lines first and see how everything is going. If the UW3-modern-english already yields (partially) meaningful results consider using this model as "starting" point (`--weights` or `--warmstart.model` parameter depending on your version).
Feel free to open a new issue if you have further obstacles.
